File: code-samples/voice/code-snippets/create-update-call-monitoring-group.cs
using System;
using System.IO;
using System.Threading.Tasks;
using System.Collections.Generic;
using RingCentral;
using Newtonsoft.Json;

namespace CreateAndManageCallMonitoringGroup {
  class Program {
    static RestClient restClient;
    static async Task Main(string[] args)
    {
      try
      {
        // Instantiate the SDK
        restClient = new RestClient("SANDBOX-APP-CLIENT-ID", "SANDBOX-APP-CLIENT-SECRET", "https://platform.devtest.ringcentral.com");

        // Authenticate a user using a personal JWT token
        await restClient.Authorize("SANDBOX-JWT");

        await create_call_monitoring_group("Demo Group - C#");
      }
      catch (Exception ex)
      {
        Console.WriteLine("Unable to authenticate to platform. Check credentials. " + ex.Message);
      }
    }
    /*
    * Create a call monitoring group
    */
    static private async Task create_call_monitoring_group(String groupName)
    {
      try
      {
        var bodyParams = new CreateCallMonitoringGroupRequest();
        bodyParams.name = groupName;
        var resp = await restClient.Restapi().Account().CallMonitoringGroups().Post(bodyParams);
        Console.WriteLine($"Call monitoring group created. Group id/name: {resp.id} / {resp.name}");
        await add_call_monitoring_group_members(resp.id);
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Unable to list call monitoring groups. {ex.Message}");
      }
    }
    /*
    * Add members to a call monitoring group
    */
    static private async Task add_call_monitoring_group_members(String groupId)
    {
      try
      {
        var newMembersList = await read_account_extensions();
        var bodyParams = new CallMonitoringBulkAssign();
        bodyParams.addedExtensions = newMembersList;
        var resp = await restClient.Restapi().Account().CallMonitoringGroups(groupId).BulkAssign().Post(bodyParams);
        Console.WriteLine("Members added.");
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Unable to add mumbers to this call monitoring group. {ex.Message}");
      }
    }
    /*
    * Read the account user extensions and create a list of supervisors and agents based on their role.
    */
    static private async Task<CallMonitoringExtensionInfo[]> read_account_extensions()
    {
      List<CallMonitoringExtensionInfo> extensionList = new List<CallMonitoringExtensionInfo>();
      var queryParams = new ListExtensionsParameters();
      queryParams.type = new String[] { "User" };
      queryParams.status = new String[] { "Enabled" };
      try
      {
        var resp = await restClient.Restapi().Account().Extension().List(queryParams);
        foreach (var record in resp.records)
        {
          CallMonitoringExtensionInfo ext = new CallMonitoringExtensionInfo();
          ext.id = record.id.ToString();
          if (record.permissions.admin.enabled == true)
          {
            ext.permissions = new String[] { "Monitoring" };
          }
          else
          {
            ext.permissions = new String[] { "Monitored" };
          }
          extensionList.Add (ext);
        }
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Unable to read account extensions. {ex.Message}");
      }
      return extensionList.ToArray();
    }
    /*
    * Delete call monitoring group
    */
    static private async Task delete_call_monitoring_group(String groupId)
    {
      try
      {
        var resp = await restClient.Restapi().Account().CallMonitoringGroups(groupId).Delete();
        Console.WriteLine("Call monitoring group deleted.");
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Unable to delete this call monitoring group. {ex.Message}");
      }
    }
  }
}

File: code-samples/voice/code-snippets/create-update-call-monitoring-group.java
package CreateAndManageCallMonitoringGroup;

import java.io.IOException;
import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;

import com.ringcentral.*;
import com.ringcentral.definitions.*;

public class CreateAndManageCallMonitoringGroup {
  static RestClient restClient;

  public static void main(String[] args) {
    var obj = new CreateAndManageCallMonitoringGroup();
    try {
      // Instantiate the SDK
      restClient = new RestClient("RC_APP_CLIENT_ID", "RC_APP_CLIENT_SECRET", "https://platform.ringcentral.com");

      // Authenticate a user using a personal JWT token
      restClient.authorize("RC_USER_JWT");

      obj.create_call_monitoring_group("Demo Group - Java");

    } catch (RestException e) {
      System.out.println(e.getMessage());
    } catch (IOException e) {
      e.printStackTrace();
    }
  }
  /*
  * Create a call monitoring group
  */
  public void create_call_monitoring_group(String groupName) throws RestException, IOException {
    try {
      var bodyParams = new CreateCallMonitoringGroupRequest();
      bodyParams.name = groupName;
      var resp = restClient.restapi().account().callMonitoringGroups().post(bodyParams);
      System.out.println("Call monitoring group created. Group id/name:" + resp.id + "/" + resp.name);
      add_call_monitoring_group_members(resp.id);
    } catch (RestException ex) {
      System.out.println("Unable to list call monitoring groups. " + ex.getMessage());
    }
  }
  /*
  * Add members to a call monitoring group
  */
  public void add_call_monitoring_group_members(String groupId) throws RestException, IOException {
    try {
      var newMembersList = read_account_extensions();
      var bodyParams = new CallMonitoringBulkAssign();
      bodyParams.addedExtensions = newMembersList;
      restClient.restapi().account().callMonitoringGroups(groupId).bulkAssign().post(bodyParams);
      System.out.println("Members added.");
    } catch (RestException ex) {
      System.out.println("Unable to add mumbers to this call monitoring group. " + ex.getMessage());
    }
  }
  /*
  * Read the account user extensions and create a list of supervisors and agents based on their role.
  */
  public CallMonitoringExtensionInfo[] read_account_extensions() throws RestException, IOException {
    ArrayList<CallMonitoringExtensionInfo> extensionList = new ArrayList<>();
    try {
      var queryParams = new ListExtensionsParameters();
      queryParams.type = new String[] { "User" };
      queryParams.status = new String[] { "Enabled" };
      var resp = restClient.restapi().account().extension().list(queryParams);
      for (var record : resp.records) {
        CallMonitoringExtensionInfo ext = new CallMonitoringExtensionInfo();
        ext.id = record.id.toString();
        if (record.permissions.admin.enabled == true)
        {
          ext.permissions = new String[] { "Monitoring" };
        }
        else
        {
          ext.permissions = new String[] { "Monitored" };
        }

        extensionList.add (ext);
      }
    } catch (RestException ex) {
      System.out.println("Unable to read account extensions. " + ex.getMessage());
    }
    CallMonitoringExtensionInfo members[]= extensionList.toArray(new CallMonitoringExtensionInfo[extensionList.size()]);
    return members;
  }
  /*
  * Delete call monitoring group
  */
  public void delete_call_monitoring_group(String groupId) throws RestException, IOException {
    try {
      restClient.restapi().account().callMonitoringGroups(groupId).delete();
      System.out.println("Call monitoring group deleted.");
    } catch (RestException ex) {
      System.out.println("Unable to delete this call monitoring group. " + ex.getMessage());
    }
  }
}
